xmodule/docs/decisions/0003-library-content-block-schema.rst

```diff
@@ -0,0 +1,201 @@
+
+We need to choose how to evolve the schema of the library_content block to support all of these incremental improvements. This is the current schema:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Field
+     - Scope, Type
+     - Default
+     - Example
+     - Description
+
+   * - source_library_id
+     - settings, str|None
+     - None
+     - "library-v1:OEX+Lib1"
+     - Key of V1 source library; None means unselected.
+
+   * - source_library\_ version
+     - settings, str|None
+     - None
+     - TBD
+     - Version of V1 source library (a MongoID string). None means latest.
+
+   * - mode
+     - settings, str
+     - "random"
+     - "random"
+     - How blocks should be selected. Only valid value is "random".
+
+   * - capa_type
+     - settings, str
+     - "any"
+     - "symbolicresponse"
+     - Unless "any": Include only problems, and only problems of this response type.
+
+   * - children
+     - special, list of XBlocks
+     - []
+     - ``block-v1:...+type@problem+block@A``
+       ``block-v1:...+type@problem+block@B``
+       ``block-v1:...+type@problem+block@D``
+       ``block-v1:...+type@problem+block@F``
+     - Copy of library contents, filtered by capa_type. Course author may apply edits that are local to this library_content block usage. Edits to child blocks' settings persist across library version upgrades. Edits to child blocks' content are wiped out when library version is upgraded.
+
+   * - max_count
+     - settings, int
+     - 1
+     - 2
+     - Number of children to randomly select for each learner. -1 means all.
+
+   * - selected
+     - user_state, list of (block type, block id) tuples
+     - []
+     - [("problem", "A"), ("problem", "F")]
+     - An ordered sub-list of the children, to be shown to a given learner. Once populated, this is kept stable for the learner, unless a the children or max_count are changed, in which case it will have blocks added or removed as necessary. If the list changes, it is re-shuffled.
+
+
+Decision
+********
+
+We will introduce the support for a "manual selection mode", without explicitly making it *mode* as hinted at by the old block schema. This will lead to a cleaner and more flexible implementation. We will achieve this with two new boolean settings: **manual** and **shuffle**:
+
+* When manual is *disabled* and shuffle is *enabled*, the block will behave as it did before. That is, it will import the entire library (filtered by capa_type) as its children, and present each user a random subset based on max_count. For backwards compatibility, these will be the default value of the settings.
+
+* When manual is *enabled*, shuffle is *disabled*, and max_count is set to -1, the block will behave in the "static" mode. That is, the user will be prompted to select specific blocks in the library, and each user will be presented those blocks in order.
```

Comment:
Consideration: At some point, I think we're going to want to be able to say "I want to select from this identified thing in the library, and potentially get updates from things I haven't selected yet" -- i.e. the equivalent of what happens to the whole library, only with a subset. Not by tags, but as a named collection–particularly if we get into a situation where we are one day mapping multiple v1 libraries into a v2 library.

Maybe we could a field that points to the library block whose children we're selecting for the course, and in the case of "whole library", it points to the root library block?

Reply:
@ormsbee Right, makes sense.

So, this?
* field name `root_block`
* settings scoped
* type is a pair of strings: `(block type, block id)`
* default is `("library", "library")` (the root library block, I think)

That begs the question for me--do V2 libraries even have "library root blocks" the way that V1 libraries do?

Reply:
@kdmccormick No, there is no "root block" in v2 libraries. It's not necessary, and we're trying to move away from "everything is an XBlock" to only "component leaves are XBlocks".

Reply:
I didn't think so. @ormsbee, I think I'm confused by your suggestion?

Reply:
> "I want to select from this identified thing in the library, and potentially get updates from things I haven't selected yet" 
Auto-updating libraries are out of scope atm, and "updates" proc on library publish, if I understand things correctly, so happy to leave this conversation for another day

Reply:
I assume that whatever field we add here would have a backwards-compatible default, so I think it's safe to leave it out of the ADR for now, and then amend the ADR if/when we decide to add it.